### PR TITLE
State

### PR DESCRIPTION
**Pull request by afeskoff-godaddy**
> [sthota](https://github.secureserver.net/sthota) **Authored by [sthota](https://github.secureserver.net/sthota)**
_<time datetime="2018-09-24T16:07:23Z" title="Monday, September 24th 2018, 6:07:23 pm +02:00">Sep 24, 2018</time>_
_Merged <time datetime="2018-09-27T20:32:05Z" title="Thursday, September 27th 2018, 10:32:05 pm +02:00">Sep 27, 2018</time>_
---

### CODE REVIEW

File: core/src/test/java/com/godaddy/commerce/core/AbstractBaseTest.java [hunk 1]
```diff
@@ -77,9 +78,67 @@ public void testValidValidateForUpdate() {
...
+    object.state(SampleState.CREATED);
 
-    Stream.of("resourceId", "restVersion", "uri", "ownerId", "version")
-      .forEach(field -> runValidateForUpdateFailure(object, field));
+    Stream.of("resourceId", "restVersion", "uri", "ownerId", "version", "state")
+        .forEach(field -> runValidateForUpdateFailure(object, field));
+  }
+
+  @Test
+  public void testStateDateSet() {
```
Comment (vdadeka-godaddy) [inline core/src/test/java/com/godaddy/commerce/core/AbstractBaseTest.java:88]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-24T16:39:23Z" title="Monday, September 24th 2018, 6:39:23 pm +02:00">Sep 24, 2018</time>_

---


You have a test case for initially setting the value (from null to non-null), and a test case for setting the value to itself. Shouldn't you also have a test case changing the value from one non-null value to another and making sure the updatedAt value changes as well?

**Review by vdadeka-godaddy — COMMENTED** [review]
> [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="undefined" title="Saturday, February 18th 2023, 2:06:52 am +01:00">Feb 18, 2023</time>_

---

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java [hunk 2]
```diff
@@ -206,7 +212,33 @@ public void validateForUpdate(AbstractBase newVersion) {
...
     newVersion.createdAt = this.createdAt;
   }
+
+  //We are giving an explicit Jackson setter method to ensure the "stateUpdatedAt" value is not accidentally set by calling the state(S) method on deserialization.
+  @JsonSetter("state")
+  private void setState(S newState) {
+    state = newState;
+  }
+
+  public void state(S newState) {
```
Comment (vdadeka-godaddy) [inline core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:229]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-24T16:40:26Z" title="Monday, September 24th 2018, 6:40:26 pm +02:00">Sep 24, 2018</time>_

---


I disagree with you removing the explicit `@JsonSetter("state")` method. I don't want to rely on Jackson to *not* pick this method when deserializing an object, and then setting the stateUpdatedAt to "now" instead of the value in the JSON.

**Review by vdadeka-godaddy — COMMENTED** [review]
> [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="undefined" title="Saturday, February 18th 2023, 2:06:56 am +01:00">Feb 18, 2023</time>_

---

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java (hunk 2 above)
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:229]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-24T16:41:41Z" title="Monday, September 24th 2018, 6:41:41 pm +02:00">Sep 24, 2018</time>_

---


And that logic will also require a test case that verifies when you deserialize a JSON object, the stateUpdatedAt() method is what was in the JSON, and not "now." I.e. set the JSON stateUpdatedAt value to something a fixed value in the past so you can verify it specifically as well.

**Review by vdadeka-godaddy — CHANGES_REQUESTED** [review]
> [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="undefined" title="Saturday, February 18th 2023, 2:07:02 am +01:00">Feb 18, 2023</time>_

---


See comments above.

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java (hunk 2 above)
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:229]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-24T16:52:19Z" title="Monday, September 24th 2018, 6:52:19 pm +02:00">Sep 24, 2018</time>_

---


And since I'm asking for what may already work without it, please put a comment on that method (which should be private, so only Jackson can see it with the annotation) that we are giving an explicit Jackson setter method to ensure the "stateUpdatedAt" value is not accidentally set by calling the `state(S)` method on deserialization.

Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:229]: > [sthota](https://github.secureserver.net/sthota) **Authored by [sthota](https://github.secureserver.net/sthota)**
_<time datetime="2018-09-24T17:06:54Z" title="Monday, September 24th 2018, 7:06:54 pm +02:00">Sep 24, 2018</time>_

---


Thank you for clearing this up.

**Comment by vdadeka-godaddy** [comment]
> [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-24T20:16:42Z" title="Monday, September 24th 2018, 10:16:42 pm +02:00">Sep 24, 2018</time>_

---


@sthota Please post this PR to #apiflowsjava. It needs a wider review at this point.

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java [hunk 3]
```diff
@@ -206,7 +212,33 @@ public void validateForUpdate(AbstractBase newVersion) {
       throw new InvalidParameterException(invalid_update, new String[]{"version"});
     }
 
+    if (state != newVersion.state) {
+      throw new InvalidParameterException(invalid_update, new String[]{"state"});
```
Comment (vdadeka-godaddy) [inline core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:216]: > [csekhar](https://github.secureserver.net/csekhar) **Authored by [csekhar](https://github.secureserver.net/csekhar)**
_<time datetime="2018-09-25T15:34:57Z" title="Tuesday, September 25th 2018, 5:34:57 pm +02:00">Sep 25, 2018</time>_

---


Why is this code in capital?

**Review by vdadeka-godaddy — COMMENTED** [review]
> [csekhar](https://github.secureserver.net/csekhar) **Authored by [csekhar](https://github.secureserver.net/csekhar)**
_<time datetime="undefined" title="Saturday, February 18th 2023, 2:19:41 am +01:00">Feb 18, 2023</time>_

---

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java [hunk 4]
```diff
@@ -19,7 +20,7 @@
 import javax.validation.constraints.Min;
 import javax.validation.constraints.Pattern;
 
-public abstract class AbstractBase implements Serializable, IResource {
+public abstract class AbstractBase<S extends Enum<S>> implements Serializable, IResource {
```
Comment (vdadeka-godaddy) [inline core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:23]: > [kcox](https://github.secureserver.net/kcox) **Authored by [kcox](https://github.secureserver.net/kcox)**
_<time datetime="2018-09-27T17:43:31Z" title="Thursday, September 27th 2018, 7:43:31 pm +02:00">Sep 27, 2018</time>_

---


Should we remove Serializable?

**Review by vdadeka-godaddy — CHANGES_REQUESTED** [review]
> [kcox](https://github.secureserver.net/kcox) **Authored by [kcox](https://github.secureserver.net/kcox)**
_<time datetime="undefined" title="Saturday, February 18th 2023, 2:19:45 am +01:00">Feb 18, 2023</time>_

---

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java [hunk 5]
```diff
@@ -206,7 +212,33 @@ public void validateForUpdate(AbstractBase newVersion) {
...
+
+  //We are giving an explicit Jackson setter method to ensure the "stateUpdatedAt" value is not accidentally set by calling the state(S) method on deserialization.
+  @JsonSetter("state")
+  private void setState(S newState) {
+    state = newState;
+  }
+
+  public void state(S newState) {
+    if (!Objects.equals(state, newState)) {
+      stateUpdatedAt = OffsetDateTime.now(ZoneOffset.UTC);
```
[the remark this answers is not in the document]
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:231]: > [kcox](https://github.secureserver.net/kcox) **Authored by [kcox](https://github.secureserver.net/kcox)**
_<time datetime="2018-09-27T18:00:27Z" title="Thursday, September 27th 2018, 8:00:27 pm +02:00">Sep 27, 2018</time>_

---


Seems like stateUpdatedAt should be internally controlled and mimic the updatedAt date.

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java (hunk 3 above)
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:216]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-27T18:05:03Z" title="Thursday, September 27th 2018, 8:05:03 pm +02:00">Sep 27, 2018</time>_

---


Per guidelines, code is uppercase with underscores : https://github.secureserver.net/CTO/guidelines/tree/master/api-design#errors

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java [hunk 6]
```diff
@@ -206,7 +212,33 @@ public void validateForUpdate(AbstractBase newVersion) {
       throw new InvalidParameterException(invalid_update, new String[]{"version"});
     }
 
+    if (!Objects.equals(state, newVersion.state)) {
```
[the remark this answers is not in the document]
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-27T18:05:51Z" title="Thursday, September 27th 2018, 8:05:51 pm +02:00">Sep 27, 2018</time>_

---


Good point - we had that like this before we made state "extends Enum". Agreed, this can be "!=".

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java [hunk 7]
```diff
@@ -206,7 +212,33 @@ public void validateForUpdate(AbstractBase newVersion) {
...
   }
+
+  //We are giving an explicit Jackson setter method to ensure the "stateUpdatedAt" value is not accidentally set by calling the state(S) method on deserialization.
+  @JsonSetter("state")
+  private void setState(S newState) {
+    state = newState;
+  }
+
+  public void state(S newState) {
+    if (!Objects.equals(state, newState)) {
```
[the remark this answers is not in the document]
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-27T18:06:59Z" title="Thursday, September 27th 2018, 8:06:59 pm +02:00">Sep 27, 2018</time>_

---


Agree again - "!=" would be good here.

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:231]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-27T18:07:28Z" title="Thursday, September 27th 2018, 8:07:28 pm +02:00">Sep 27, 2018</time>_

---


I thought we discussed this at the architecture meeting, and decided not to worry about it matching exactly.

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java (hunk 4 above)
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:23]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-27T18:26:40Z" title="Thursday, September 27th 2018, 8:26:40 pm +02:00">Sep 27, 2018</time>_

---


I remember we discussed this previously. Sri - if you simply remove Serializable from this class, and rerun through Sonar, does it trigger any bugs/issues?

Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:23]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-27T18:27:02Z" title="Thursday, September 27th 2018, 8:27:02 pm +02:00">Sep 27, 2018</time>_

---


BTW, that request is separate from this update. I'm not sure we should extend the scope of this PR to tackle that prior request.

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java [hunk 8]
```diff
@@ -206,7 +212,33 @@ public void validateForUpdate(AbstractBase newVersion) {
       throw new InvalidParameterException(invalid_update, new String[]{"version"});
     }
 
+    if (!Objects.equals(state, newVersion.state)) {
+      throw new InvalidParameterException("STATE_CHANGE", new String[]{"state"});
```
[the remark this answers is not in the document]
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-27T18:28:04Z" title="Thursday, September 27th 2018, 8:28:04 pm +02:00">Sep 27, 2018</time>_

---


We aren't including old/new values in the other exceptions. I wouldn't add it here either.

[the remark this answers is not in the document]
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java]: > [kcox](https://github.secureserver.net/kcox) **Authored by [kcox](https://github.secureserver.net/kcox)**
_<time datetime="2018-09-27T18:36:48Z" title="Thursday, September 27th 2018, 8:36:48 pm +02:00">Sep 27, 2018</time>_

---


What is value of String[]{"state"} when STATE_CHANGE is the message?

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:231]: > [kcox](https://github.secureserver.net/kcox) **Authored by [kcox](https://github.secureserver.net/kcox)**
_<time datetime="2018-09-27T18:39:43Z" title="Thursday, September 27th 2018, 8:39:43 pm +02:00">Sep 27, 2018</time>_

---


Seems like it would be easy to correlate the two here rather than letting them drift. If it's hard, then let it go... I thought was to have a single private internal only date and set both dates from it.

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-27T18:40:19Z" title="Thursday, September 27th 2018, 8:40:19 pm +02:00">Sep 27, 2018</time>_

---


Ahh, I see... This should actually agree with the other throws above. @sthota it should also throw "invalid_update", and then "state" stays in the String array.

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java (hunk 4 above)
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:23]: > [sthota](https://github.secureserver.net/sthota) **Authored by [sthota](https://github.secureserver.net/sthota)**
_<time datetime="2018-09-27T18:41:11Z" title="Thursday, September 27th 2018, 8:41:11 pm +02:00">Sep 27, 2018</time>_

---


We should leave it for another PR. I thought of removing it in the PR, but didn't want unrelated updates

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:231]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-27T18:48:10Z" title="Thursday, September 27th 2018, 8:48:10 pm +02:00">Sep 27, 2018</time>_

---


That is an interesting point. We already have a time-sync issue between createdAt and updatedAt. This introduces a third date to the JSON which can have a similar minor variance. Maybe we could address all three of these dates in a separate update (if it turns out to be a problem to our users).

The easiest way I see doing this would be having private boolean flags indicating which dates we want to set when the Store calls `updatedAt(OffsetDateTime updatedAt)` to update one or more dates. Which isn't a big issue, but maybe we could put a pin in that request and fix both timestamps in an update.

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java [hunk 9]
```diff
@@ -206,7 +212,33 @@ public void validateForUpdate(AbstractBase newVersion) {
       throw new InvalidParameterException(invalid_update, new String[]{"version"});
     }
 
+    if (state != newVersion.state) {
+      throw new InvalidParameterException("STATE_CHANGE", "Cannot update state");
```
Comment (vdadeka-godaddy) [inline core/src/main/java/com/godaddy/commerce/model/AbstractBase.java]: > [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="2018-09-27T18:50:04Z" title="Thursday, September 27th 2018, 8:50:04 pm +02:00">Sep 27, 2018</time>_

---


This should be

`throw new InvalidParameterException(invalid_update, new String[]{"state"});`

to agree with the InvalidParameterExceptions above this one.

**Review by vdadeka-godaddy — COMMENTED** [review]
> [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="undefined" title="Saturday, February 18th 2023, 2:24:58 am +01:00">Feb 18, 2023</time>_

---

File: core/src/main/java/com/godaddy/commerce/model/AbstractBase.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vdadeka-godaddy) [reply core/src/main/java/com/godaddy/commerce/model/AbstractBase.java:231]: > [sthota](https://github.secureserver.net/sthota) **Authored by [sthota](https://github.secureserver.net/sthota)**
_<time datetime="2018-09-27T18:53:36Z" title="Thursday, September 27th 2018, 8:53:36 pm +02:00">Sep 27, 2018</time>_

---


There is a similar issue that was created regarding createdAt/ updatedAt timestamp sync.  We should address all these 3 at some point.

**Review by vdadeka-godaddy — APPROVED** [review]
> [lhostetler](https://github.secureserver.net/lhostetler) **Authored by [lhostetler](https://github.secureserver.net/lhostetler)**
_<time datetime="undefined" title="Saturday, February 18th 2023, 2:25:05 am +01:00">Feb 18, 2023</time>_

---


:+1:

**Review by vdadeka-godaddy — APPROVED** [review]
> [kcox](https://github.secureserver.net/kcox) **Authored by [kcox](https://github.secureserver.net/kcox)**
_<time datetime="undefined" title="Saturday, February 18th 2023, 2:25:13 am +01:00">Feb 18, 2023</time>_

---
